fms/models/mixtral.py

```diff
@@ -0,0 +1,435 @@
+        # x: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [
+                (torch.empty(0), torch.empty(0)) for _ in range(len(self.layers))
+            ]
+
+        qlen = x.size(1)
+        klen = x.size(1)
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0][0].numel() > 0:
+            klen += past_key_value_states[0][0].size(1)
+
+        # if mask is none, we need to specify causal mask
+        if mask is None:
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                is_causal_mask = False
+            else:
+                is_causal_mask = True
+        else:
+            is_causal_mask = False
+
+        x = self.embedding(x)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            output = layer(
+                x=x,
+                mask=mask,
+                position_ids=position_ids,
+                past_key_value_state=past_key_value_states[i],
+                use_cache=use_cache,
+                is_causal_mask=is_causal_mask,
+                attn_algorithm=attn_algorithm,
+            )
+
+            if use_cache:
+                x, present_key_value_state = output
+                present_key_value_states.append(present_key_value_state)
+
+            else:
+                x = output
+
+        dec_out = x
+        dec_out = self.dec_norm(dec_out)
+        if self.config.p_dropout:
+            dec_out = self.dropout(dec_out)
+
+        return dec_out, present_key_value_states
+
+
+class Mixtral(nn.Module):
+    def __init__(
+        self,
+        config: Optional[MixtralConfig] = None,
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+        **kwargs,
+    ):
+        super(Mixtral, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = MixtralConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        self.base_model = MixtralHeadless(self.config, self.distributed_strategy)
+        head = LMHead(self.config.dim, self.config.src_vocab_size, bias=False)
+        self.head = self.distributed_strategy.distribute_module(head)
+
+        self.reset_params()
+
+    def get_config(self) -> MixtralConfig:
+        return self.config
+
+    @classmethod
+    def from_config(cls, config: MixtralConfig) -> "Mixtral":
+        return cls(config)
+
+    def reset_params(self):
```

Comment:
i believe this should be `reset_parameters` not `reset_params`. (that's what we use in llama and what we check for in get_model)